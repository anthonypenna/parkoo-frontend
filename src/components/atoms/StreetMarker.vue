<template>
  <svg
    width="32"
    height="48"
    viewBox="0 0 32 48"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
    :style="`filter: drop-shadow(0 4px 16px ${color}A0)`"
  >
    <path
      fill-rule="evenodd"
      clip-rule="evenodd"
      d="M32 16C32 19.014 31.1666 21.8333 29.7178 24.2401L16 48L2.28223 24.2401C0.833362 21.8333 0 19.014 0 16C0 7.16344 7.16344 0 16 0C24.8366 0 32 7.16344 32 16ZM16 24C20.4183 24 24 20.4183 24 16C24 11.5817 20.4183 8 16 8C11.5817 8 8 11.5817 8 16C8 20.4183 11.5817 24 16 24Z"
      :fill="color"
    />
  </svg>
</template>

<script lang="ts">
import Vue from "vue";

export default Vue.extend({
  name: "StreetMarker",

  props: {
    cleanedTomorrow: {
      default: false,
      type: Boolean,
    },
  },

  computed: {
    color() {
      return this.cleanedTomorrow ? "#FF7373" : "#1DDC3C";
    },
  },
});
</script>
