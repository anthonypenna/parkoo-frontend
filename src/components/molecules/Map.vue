<template>
  <div :id="id" style="width: 100%; height: 100vh">
    <slot />
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import { mapGetters, mapMutations } from "vuex";
import * as Mapbox from "mapbox-gl";
import { MAPBOX_CENTER, MAPBOX_THEME, MAPBOX_ZOOM } from "@/constants/mapbox";
import { MapChild } from "@/models/MapChild";

export default Vue.extend({
  name: "Map",

  props: {
    id: {
      default: "map",
      type: String,
    },

    accessToken: {
      default: "",
      type: String,
    },

    center: {
      default: () => MAPBOX_CENTER,
      type: Array,
    },

    theme: {
      default: MAPBOX_THEME,
      type: String,
    },

    zoom: {
      default: MAPBOX_ZOOM,
      type: Number,
    },
  },

  computed: {
    ...mapGetters("mapbox", ["map"]),
  },

  methods: {
    ...mapMutations("mapbox", ["setMap"]),

    renderChildComponent<A extends MapChild>(childComponent: A | undefined) {
      if (childComponent) {
        const { $el, lat, lng } = childComponent;
        const element = $el as HTMLElement;
        new Mapbox.Marker({ element }).setLngLat([lng, lat]).addTo(this.map);
      }
    },

    onMapLoad() {
      this.$children.forEach((child) =>
        this.renderChildComponent(child as MapChild)
      );
      this.$emit("load");
    },

    onMapMoveEnd() {
      this.$emit("moveend");
    },
  },

  mounted() {
    const map = new Mapbox.Map({
      container: this.id,
      accessToken: this.accessToken,
      center: this.center as Mapbox.LngLatLike,
      style: this.theme,
      zoom: this.zoom,
    });

    this.setMap(map);
    map.on("load", this.onMapLoad);
  },
});
</script>
