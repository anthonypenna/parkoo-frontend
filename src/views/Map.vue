<template>
  <div class="map">
    <ParkooMap
      :accessToken="accessToken"
      :center="center"
      :theme="theme"
      :zoom="zoom"
      @load="onMapLoad"
    >
      <UserMarker :lat="lat" :lng="lng" />
      <StreetMarker
        v-for="street of streets"
        :key="street._id"
        :lat="street.lat"
        :lng="street.lng"
        :cleanedTomorrow="isStreetCleanedTomorrow(street)"
      />
    </ParkooMap>
    <NoStreetsModal
      v-show="showNoStreetsModal"
      @close="onNoStreetsModalClose"
      @addstreet="onNoStreetsModalAddStreet"
    />
    <AddStreetModal
      v-show="showAddStreetModal"
      @cancel="onAddStreetModalCancel"
      @add="onAddStreetModalAdd"
    />
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import ParkooMap from "@/components/molecules/Map.vue";
import UserMarker from "@/components/atoms/UserCursor.vue";
import StreetMarker from "@/components/atoms/StreetMarker.vue";
import NoStreetsModal from "@/components/molecules/NoStreetsModal.vue";
import AddStreetModal from "@/components/molecules/AddStreetModal.vue";
import { mapGetters, mapMutations } from "vuex";
import { LngLat, Map } from "mapbox-gl";
import { areEqual } from "@/utils/array";
import { MAPBOX_FLY_SPEED } from "@/constants/mapbox";
import { mapMiddleware } from "@/middleware/map";

export default Vue.extend({
  name: "MapView",
  mixins: [mapMiddleware],

  components: {
    ParkooMap,
    UserMarker,
    StreetMarker,
    NoStreetsModal,
    AddStreetModal,
  },

  computed: {
    ...mapGetters("mapbox", ["accessToken", "center", "map", "theme", "zoom"]),
    ...mapGetters("user", ["lat", "lng"]),
    ...mapGetters("streets", [
      "streets",
      "hasStreets",
      "isStreetCleanedTomorrow",
    ]),
    ...mapGetters(["showNoStreetsModal", "showAddStreetModal"]),
  },

  methods: {
    ...mapMutations([
      "setShowLoading",
      "setShowNoStreetsModal",
      "setShowAddStreetModal",
    ]),

    onMapLoad() {
      const map = this.map as Map;
      const mapCenter = map.getCenter().toArray();
      const userPosition = new LngLat(this.lng, this.lat);

      if (areEqual(mapCenter, userPosition.toArray())) {
        this.onMapReady();
        return;
      }

      map.once("moveend", this.onMapReady);
      map.flyTo({ center: userPosition, speed: MAPBOX_FLY_SPEED });
    },

    onMapReady() {
      this.setShowLoading(false);
      if (!this.hasStreets) this.setShowNoStreetsModal(true);
    },

    onNoStreetsModalClose() {
      this.setShowNoStreetsModal(false);
    },

    onNoStreetsModalAddStreet() {
      this.setShowAddStreetModal(true);
    },

    onAddStreetModalCancel() {
      this.setShowAddStreetModal(false);
    },

    onAddStreetModalAdd() {
      alert("Todo!");
    },
  },
});
</script>
