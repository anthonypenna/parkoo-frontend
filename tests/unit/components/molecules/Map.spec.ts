import { createLocalVue, mount, Wrapper } from '@vue/test-utils'
import Vuex from 'vuex'
import Map from '@/components/molecules/Map.vue'
import StreetMarker from '@/components/atoms/StreetMarker.vue'
import * as Mapbox from 'mapbox-gl'

type MapComponent = Wrapper<
  Map & {
    map: Mapbox.Map
    renderChildComponent: (component: unknown) => unknown
    onMapLoad: () => void
  },
  Element
>

jest.mock('mapbox-gl', () => ({
  Map: jest.fn(() => ({ on: jest.fn() })),
  Marker: jest.fn(() => ({
    setLngLat: jest.fn().mockReturnValue({ addTo: jest.fn() }),
  })),
}))

const localVue = createLocalVue()

localVue.use(Vuex)

const storeOptions = {
  modules: {
    mapbox: {
      namespaced: true,
      state: {
        map: null,
      },
      mutations: {
        setMap: jest.fn(),
      },
      getters: {
        map: jest.fn(),
      },
    },
  },
}

const store = new Vuex.Store(storeOptions)

const propsData = {
  accessToken: 'token',
  center: [0, 0],
  theme: 'theme',
  zoom: 1,
}

describe('<Map />', () => {
  const initialState = JSON.stringify(store.state)

  beforeEach(() => {
    store.replaceState(JSON.parse(initialState))
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should render correctly', () => {
    const wrapper = mount(Map, { localVue, propsData, store })
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('when mounted', () => {
    it('should create a new mapbox map', () => {
      mount(Map, { localVue, propsData, store })
      expect(Mapbox.Map).toHaveBeenCalledWith({
        container: 'map',
        accessToken: propsData.accessToken,
        center: propsData.center,
        style: propsData.theme,
        zoom: propsData.zoom,
      })
    })

    it('should save the map instance to the store', () => {
      mount(Map, { localVue, propsData, store }) as MapComponent
      expect(storeOptions.modules.mapbox.mutations.setMap).toHaveBeenCalled()
    })
  })

  describe('methods', () => {
    describe('renderChildComponent', () => {
      describe('when the component is undefined', () => {
        it('should do nothing', () => {
          const wrapper = mount(Map, {
            localVue,
            propsData,
            store,
          }) as MapComponent

          wrapper.vm.renderChildComponent(undefined)
          expect(Mapbox.Marker).not.toHaveBeenCalled()
        })
      })

      describe('when the component is defined', () => {
        it('should add a new marker to the map', () => {
          const wrapper = mount(Map, {
            localVue,
            propsData,
            store,
          }) as MapComponent

          const component = { $el: {}, lat: 45, lng: 9 }
          wrapper.vm.renderChildComponent(component)
          expect(Mapbox.Marker).toHaveBeenCalledWith({ element: component.$el })
        })
      })
    })

    describe('onMapLoad', () => {
      it('should render every child component', () => {
        const wrapper = mount(Map, {
          components: { StreetMarker },
          localVue,
          propsData,
          store,
          slots: {
            default: `
                <StreetMarker />
                <StreetMarker />
                <StreetMarker />
          `,
          },
        }) as MapComponent

        wrapper.vm.renderChildComponent = jest.fn()
        wrapper.vm.onMapLoad()
        expect(wrapper.vm.renderChildComponent).toHaveBeenCalledTimes(3)
      })

      it('should emit a load event', () => {
        const wrapper = mount(Map, {
          localVue,
          propsData,
          store,
        }) as MapComponent

        wrapper.vm.onMapLoad()
        expect(wrapper.emitted().load).toBeTruthy()
      })
    })
  })
})
