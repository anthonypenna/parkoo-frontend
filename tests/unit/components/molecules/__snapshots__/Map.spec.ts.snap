// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> should render correctly 1`] = `<div id="map" style="width: 100%; height: 100vh;"></div>`;
